pkg/standalone-utils/test/LidoRelayer.test.ts

```diff
@@ -72,383 +65,543 @@ describe('LidoRelayer', function () {
-        fromInternalBalance: false,
-        recipient: sender.address,
-        toInternalBalance: false,
-      };
-    });
-
-    context('when the relayer is authorized', () => {
-      sharedBeforeEach('allow relayer', async () => {
-        const manageUserBalanceAction = await actionId(vault.instance, 'manageUserBalance');
-        const swapAction = await actionId(vault.instance, 'swap');
-
-        await vault.authorizer?.connect(admin).grantRoles([manageUserBalanceAction, swapAction], relayer.address);
-      });
-
-      context('when the user did allow the relayer', () => {
-        sharedBeforeEach('allow relayer', async () => {
-          await vault.instance.connect(sender).setRelayerApproval(sender.address, relayer.address, true);
-        });
+  sharedBeforeEach('set up relayer', async () => {
+    // Deploy Relayer
+    relayerLibrary = await deploy('LidoBatchRelayerLibrary', { args: [vault.address, wstETH.address] });
+    relayer = await deployedAt('BalancerRelayer', await relayerLibrary.getEntrypoint());
+
+    // Authorize Relayer for all actions
+    const relayerActionIds = await Promise.all(
+      ['swap', 'batchSwap', 'joinPool', 'exitPool', 'setRelayerApproval', 'manageUserBalance'].map((action) =>
+        actionId(vault.instance, action)
+      )
+    );
+    const authorizer = await deployedAt('v2-vault/Authorizer', await vault.instance.getAuthorizer());
+    await authorizer.connect(admin).grantRoles(relayerActionIds, relayer.address);
+
+    // Approve relayer by sender
+    await vault.instance.connect(sender).setRelayerApproval(sender.address, relayer.address, true);
+  });
 
-        describe('swap using stETH as an input', () => {
-          let singleSwap: SingleSwap;
+  const CHAINED_REFERENCE_PREFIX = 'ba10';
+  function toChainedReference(key: BigNumberish): BigNumber {
+    // The full padded prefix is 66 characters long, with 64 hex characters and the 0x prefix.
+    const paddedPrefix = `0x${CHAINED_REFERENCE_PREFIX}${'0'.repeat(64 - CHAINED_REFERENCE_PREFIX.length)}`;
+
+    return BigNumber.from(paddedPrefix).add(key);
+  }
+
+  function encodeApprove(token: Token, amount: BigNumberish): string {
+    return relayerLibrary.interface.encodeFunctionData('approveVault', [token.address, amount]);
+  }
+
+  function encodeWrap(sender: string, recipient: string, amount: BigNumberish, outputReference?: BigNumberish): string {
+    return relayerLibrary.interface.encodeFunctionData('wrapStETH', [sender, recipient, amount, outputReference ?? 0]);
+  }
+
+  function encodeUnwrap(
+    sender: string,
+    recipient: string,
+    amount: BigNumberish,
+    outputReference?: BigNumberish
+  ): string {
+    return relayerLibrary.interface.encodeFunctionData('unwrapWstETH', [
+      sender,
+      recipient,
+      amount,
+      outputReference ?? 0,
+    ]);
+  }
 
-          sharedBeforeEach('build swap request', async () => {
-            singleSwap = {
+  describe('swap', () => {
```

Comment:
These tests were good back when the primitives of the relayer were swap, join, etc. I'd change this up now for more focused tests the use the different functionalities of the wrapper library (chained references, recipient redirection, etc.)

Reply:
I've started adding these tests for primitives although I think it's helpful to have some more complex tests in cases where we have a well defined expected usecase for the module being tested (e.g. for `LidoWrapping` it would be wrap steth and trade using wstETH)

Reply:
Yes, testing the expected use cases makes sense as that may reveal things we didn't notice (such as the approval for wstETH that was mentioned in some other comment). But I still believe a better balance would be to have full unit test style coverage of the primitives, and then a few 'integration' tests at the end (like we do when chaining swaps and joins).